[BEAM-5910] Add lastModified field to MatchResult.Metadata

jklukas opened the pull request:
In the Java SDK, the Filesystems.match facilities are aimed primarily at listing file names and collect very limited additional metadata from the filesystem (sizeBytes and isReadSeekEfficient). This PR adds a new `lastModified` field to that list.

This could be a basis for a future improvement to FileIO.match(...).continuously(...) where we could let the user opt to poll not just for new file names, but also for existing file names if their content has been updated.

In the near term, the addition of lastModified to Metadata will allow users to implement their own polling logic on top of Filesystems.match to detect and download new files from any of the supported filesystems.


------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

jklukas commented:
I'm expecting some tests to fail initially. Seeing what fails will let me know what tests need to be updated.

jklukas commented:
retest this please

jklukas commented:
retest this please

iemejia commented:
Just for comment, the PR looks great, but waiting for the discussion on @dev to see what is the recommended path for the MetadataCoder update.

jklukas commented:
Thanks for reviewing, @iemejia. I'm very interested to see what comes out of the discussion on evolving coders.

iemejia commented:
Yes hopefully it won't take long, but probably will require a rebase (assuming that we will version the coders and save this somewhere).

jklukas commented:
@iemejia - [The mailing list discussion](https://lists.apache.org/thread.html/d9a780bd5203b5188168d6fb461e6822dcac4314fd4530873876ec3c@%3Cdev.beam.apache.org%3E) looks to have reached a conclusion that there's no short-term solution for coder versioning, so we need to be conservative about compatibility.

I've added a new commit here that returns `MetadataCoder` to compatibility with the existing format, providing a default value for lastModifiedMillis. We introduce a `MetadataCoderV2` that encodes lastModifiedMillis, but is strictly opt-in; it's not used directly by the SDK at all.

iemejia commented:
Sure @jklukas sorry for the delay will take a look ASAP (Just back from christmass holidays).

iemejia commented:
Oups just realized I reviewed the wrong one hehe never mind, will start to check this one now. Sorry.

iemejia commented:
Sorry for the delay @jklukas was a bit busy with other stuff + finishing other reviews, your PR is next in the line.

iemejia commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/fs/MetadataCoder.java:
```diff
@@ -26,8 +26,16 @@
+ * MetadataCoderV2} for retaining timestamp information.
+ */
 public class MetadataCoder extends AtomicCoder<Metadata> {
+  public static final long UNKNOWN_LAST_MODIFIED_MILLIS = -1L;
```
Remove see comment below

iemejia commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/fs/MetadataCoderV2.java:
```diff
@@ -0,0 +1,58 @@
+  private static final VarLongCoder LONG_CODER = VarLongCoder.of();
+
+  /** Creates a {@link MetadataCoder}. */
+  public static MetadataCoder of() {
```
This one should return `MetadataCoderV2`.

iemejia commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/fs/MetadataCoderV2.java:
```diff
@@ -0,0 +1,58 @@
+
+  /** Creates a {@link MetadataCoder}. */
+  public static MetadataCoder of() {
+    return new MetadataCoder();
```
Can you please make this one a singleton. Optionally can you do the same for the original `MetadataCoder` (I just discover it was also building an instance per use). See `TableDestinationCoder` for the singleton style aka private constructor + private static INSTANCE pattern on Beam coders.

jklukas commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/fs/MetadataCoder.java:
```diff
@@ -46,14 +54,18 @@ public void encode(Metadata value, OutputStream os) throws IOException {
         .setIsReadSeekEfficient(isReadSeekEfficient)
         .setSizeBytes(sizeBytes)
-        .build();
+        .setLastModifiedMillis(UNKNOWN_LAST_MODIFIED_MILLIS);
```
The idea here is that lastModifiedMillis must have some default value. I've chosen here to use `-1` so that `lastModifiedMillis` doesn't have to be nullable, though we can discuss if null makes more sense here.  

AutoValue requires that values be set for all members. To define a default value, the AutoValue docs suggest a pattern of calling setters before returning the builder. That's what's going on here.

I don't see that there's a backwards incompatibility here. MetadataCoder still only writes and reads the three existing values (resource id, int, and long). When encoding, it throws away lastModifiedMills and when decoding, it provides the default -1 value.

Do you have thoughts on whether `null` would be preferred as the default vs. `-1`?

jklukas replied:
Of note, Java's `File.lastModified` documents that `0` is used if the file does not exist or an I/O error occurs. We could follow that and use `0`.

iemejia replied:
What if we let zero as a default. In that case we won't need that and we will still be ok with AutoValue. If we setup any default I think it is better to do it at the core object level `Metadata.create()`  better than at other places. Also in the current case if you put this only in `decode` then encode won't put the right value. Better to left this class as it was to avoid the risk of breaking stuff..

jklukas replied:
> What if we let zero as a default. In that case we won't need that and we will still be ok with AutoValue

To be clear, `0` is only used as a default for `File.lastModified` as used in `LocalFileSystem`. Other filesystems do not have this same concept of 0 as default. So any way we go, we will have to explicitly set some default value for the field.

I agree with you that it's better to set the default within `Metadata` rather than only in the coder, so I will make that change. And will also change the default to be 0 to be consistent with `File.lastModified` behavior.

> Also in the current case if you put this only in decode then encode won't put the right value

Exactly. `encode` _does not_ encode the lastModifiedMillis value; it's thrown away. And decode provides a default. So encode and decode remain matched in the number and types of encoded values, and these are compatible with previous beam versions.

jklukas commented:
Fixed the typos, converted both MetadataCoder and MetadataCoderV2 to singletons, and added tests.

I responded to the comment on `UNKNOWN_LAST_MODIFIED_MILLIS`. The tests exercise that the new code and validate that behavior is how I intended (MetadataCoder discards lastModifiedMillis when encoding and provides a default value on decoding). I believe that proves backwards compatibility, but do let me know if there's some additional nuance I'm missing there.

jklukas commented:
Pushed a new commit that uses `0` for the default, moves the default into `Metadata`, and documents the behavior.
